### ISSUE TITLE
Checking for inconsistencies in parsers and writers methods

### ISSUE THREAD

**Issue by javihern98** [bis-med-it/pysdmx#164]
Check all parsers and writers follow these guidelines:

- Same arguments order: data, path, (modifiers, mapping, etc)
- Same name for same functions: 
  - Public method names must be "read" and "write"
  - Private and local methods will have _ or __ respectively
- Input data/structures must be a dictionary
  - Typing must be defined as precise as possible, avoiding Any
  - Allow for several datasets writing on each method
  - Add checks to avoid any possible misuse of the methods (example: pass an ItemScheme to a data writer, or pass a Dataset to structures writing)

**Comment by javihern98** [comment]
@sosna Let´s add here all guidelines for parsers and writers and make the changes, please comment if any other guideline is necessary

**Comment by javihern98** [comment]
Agreed on meeting to sepatare data readers and writers and metadata read and writers for SDMX-ML

**Comment by sosna** [comment]
Having a look at the readers, it looks like some functionality is also format-specific? For example, the XML readers offer a boolean flag called `validate`, while the CSV readers don't. 

I guess it might OK but it can come as a surprise to some that such a feature is format-specific...

**Comment by sosna** [comment]
Thanks, @javihern98. What do you mean exactly with CSV being always validated?

Is it a kind of syntax validation (i.e. CSV is valid with [RFC 4180](https://www.ietf.org/rfc/rfc4180.txt)). Or did you mean semantic validation (i.e. against the datastructure-, dataflow- or provisionagreement-related schema)?

**Comment by javihern98** [comment]
> Thanks, [@javihern98](https://github.com/javihern98). What do you mean exactly with CSV being always validated?
> 
> Is it a kind of syntax validation (i.e. CSV is valid with [RFC 4180](https://www.ietf.org/rfc/rfc4180.txt)). Or did you mean semantic validation (i.e. against the datastructure-, dataflow- or provisionagreement-related schema)?

I refer only to structural validation, i.e. to be compliant with RFC 4180 and the SDMX-CSV specification

**Comment by sosna** [comment]
OK, thanks. For clarification purposes, maybe we should distinguish between:

- Syntax validation (i.e. is the message well-formed?)
- Structure validation (i.e. is the message following the expected "template"?)
- Semantic validation (i.e. is the message compliant with the expected structural metadata?)

I guess that syntax validation comes out of the box, i.e. if a message is invalid (for example, an XML element is left unclosed), then the parser will throw an error.

For structure validation, it's merely checking that the reader finds the structure it expects. For example, in SDMX-CSV, does it start with the expected special columns?

Under semantic validation, we would cover the compliance with the structural metadata (data structure, dataflow, provision agreement, etc.). 

In SDMX-ML, types 2 and 3 are currently checked using XML schemas. But we already know that this is not sufficient, i.e. there are various categories of errors that cannot be checked with an XML schema.

So, for the future, could we agree that, if the `validate` flag is set to `false`, types 1 and 2 will be checked, while, if `validate` is set to `true`, then all 3 types of validations will be performed? 

These features would need to be supported regardless of the format and so, even if we don't add such validation now, we should at least make sure the API is the same across formats. Would you agree?

**Comment by mla2001** [comment]
I have a question about one of the parameters of the old XML reader. It is 'mode', which is apparently only used in the following part:

```
if mode is not None and MODES[mode.value] not in dict_info:
    raise Invalid(
        "Validation Error",
        f"Unable to parse sdmx file as {MODES[mode.value]} file",
    )
```

So... It is only used to see that it is not in the dictionary of modes, but nothing more, then it is not passed as a parameter to any method or used in any other sentence. Should I delete it in the new reader or does it have any future use?

**Comment by javihern98** [comment]
Agreed on meeting to change the validate flag for a better term (maybe schema_validation?)

**Comment by javihern98** [comment]
> I have a question about one of the parameters of the old XML reader. It is 'mode', which is apparently only used in the following part:
> 
> ```
> if mode is not None and MODES[mode.value] not in dict_info:
>     raise Invalid(
>         "Validation Error",
>         f"Unable to parse sdmx file as {MODES[mode.value]} file",
>     )
> ```
> 
> So... It is only used to see that it is not in the dictionary of modes, but nothing more, then it is not passed as a parameter to any method or used in any other sentence. Should I delete it in the new reader or does it have any future use?

Yes, the mode should be removed if we separate the readers for data and metadata on SDMX-ML

**Comment by mla2001** [comment]
I have a doubt, would there be a way to infer in the case of data whether the message type would be GenericDataset or StructureSpecificdataset?
To continue with the implementation of two writers (data and metadata) with the proposed prototype:
```
def writer(datasets: Sequence[PandasDataset], output_path: Optional[str] = None, prettyprint: bool = True, header: Optional[Header] = None, dimension_at_observation: Optional[Dict[str, str]] = None):
    Pass
```
If this is not possible, we would have to continue with the other option, a single writer in which the type of message is indicated.
```
def writer(datasets: Sequence[PandasDataset], output_path: Optional[str] = None, type_: MessageType = MessageType.StructureSpecificDataSet, prettyprint: bool = True, header: Optional[Header] = None, dimension_at_observation: Optional[Dict[str, str]] = None):
    Pass
```

**Comment by javihern98** [comment]
> I have a doubt, would there be a way to infer in the case of data whether the message type would be GenericDataset or StructureSpecificdataset? To continue with the implementation of two writers (data and metadata) with the proposed prototype:
> 
> ```
> def writer(datasets: Sequence[PandasDataset], output_path: Optional[str] = None, prettyprint: bool = True, header: Optional[Header] = None, dimension_at_observation: Optional[Dict[str, str]] = None):
>     Pass
> ```
> 
> If this is not possible, we would have to continue with the other option, a single writer in which the type of message is indicated.
> 
> ```
> def writer(datasets: Sequence[PandasDataset], output_path: Optional[str] = None, type_: MessageType = MessageType.StructureSpecificDataSet, prettyprint: bool = True, header: Optional[Header] = None, dimension_at_observation: Optional[Dict[str, str]] = None):
>     Pass
> ```

Agreed on meeting:

- A specific writer for each SDMX-ML message type will be generated, with proper signature:
  - data: Sequence[Dataset], dimension_at_observation: Optional[Dict[str, str]] = None
  - Structures: Dict[str, Dict[...]]
  - (SubmissionResult and Error does not have a writer implemented)
  - These writers will have a Header and prettyprint parameters at the end.

- We must keep consistency between SDMX-ML and SDMX-CSV on data writer, the data must be always the first parameter and the output_path must be always the second parameter.

**Comment by javihern98** [comment]
@sosna @stratosn @antonio-olleros We currently have as structures result and input of the writer a dictionary with this typing:
```python
Dict[
            str,
            Dict[
                str,
                Union[
                    ItemScheme,
                    Codelist,
                    ConceptScheme,
                    DataStructureDefinition,
                    Dataflow,
                ],
            ],
        ]
```
Which I believe its too complex. Shall we simplify it to this?:

```python
Sequence[
               Union[
                    ItemScheme,
                    Codelist,
                    ConceptScheme,
                    DataStructureDefinition,
                    Dataflow,
                ],
]
```

**Comment by sosna** [comment]
This sounds nicer to me as well. Thanks, @javihern98

**Comment by sosna** [comment]
> * These writers will have a Header and prettyprint parameters at the end. StructureSpecific and Generic wirters must have the same signature.

IMHO, it is important to offer readers for both StructureSpecific and Generic data messages. However, for writing, it might be sufficient to support StructureSpecific data messages only. After all, Generic data messages have been deprecated since SDMX 3.0 was released in 2021.

`pysdmx` is based on SDMX 3.0 (the model, the API, etc.) and so, in general, I think it's OK if we don't write messages in formats that are no longer available (like the SDMX-ML Cross-Sectional Data messages and the SDMX-ML Utility Data Messages, both deprecated in 2011 with the release of SDMX 2.1, and the SDMX-ML Generic Data message, deprecated in 2021 with the release of SDMX 3.0).

Just my 2 cents :)

**Comment by javihern98** [comment]
> > * These writers will have a Header and prettyprint parameters at the end. StructureSpecific and Generic wirters must have the same signature.
> 
> IMHO, it is important to offer readers for both StructureSpecific and Generic data messages. However, for writing, it might be sufficient to support StructureSpecific data messages only. After all, Generic data messages have been deprecated since SDMX 3.0 was released in 2021.
> 
> `pysdmx` is based on SDMX 3.0 (the model, the API, etc.) and so, in general, I think it's OK if we don't write messages in formats that are no longer available (like the SDMX-ML Cross-Sectional Data messages and the SDMX-ML Utility Data Messages, both deprecated in 2011 with the release of SDMX 2.1, and the SDMX-ML Generic Data message, deprecated in 2021 with the release of SDMX 3.0).
> 
> The comment above applies to **writers** only! Ideally, we should have readers for every SDMX formats that ever existed 😁.
> 
> Just my 2 cents :)

Re. the generic format, as it is already implemented we can keep the writer, just changing the signature to adapt it accordingly
